PXC Stability: Fixed PXC-3446, PXC-3552 and PXC-3553

venkatesh-prasad-v opened the pull request:
1. PXC-3446 - Memory leak during server shutdown
    
    https://jira.percona.com/browse/PXC-3446
    
    Problem: Almost every MTR test exposed the following memory leaks when
    run in cluster mode.
    
    1. Memory leak in rollbacker thread arguments.
    
    ==111465== 16 bytes in 1 blocks are definitely lost in loss record 2 of 4
    ==111465==    at 0x89ECC83: operator new(unsigned long)
    ==111465==    by 0x389DC9A: wsrep_create_rollbacker() (wsrep_thd.cc:309)
    ==111465==    by 0x3874FBE: wsrep_init_startup(bool) (wsrep_mysqld.cc:1125)
    ==111465==    by 0x34010C5: init_server_components() (mysqld.cc:6199)
    ==111465==    by 0x340715B: mysqld_main(int, char**) (mysqld.cc:7689)
    ==111465==    by 0x31A9919: main (main.cc:25)
    
    2. Memory leak in `wsrep_cluster_name` server variable.
    
    ==111465== 49 bytes in 1 blocks are definitely lost in loss record 4 of 4
    ==111465==    at 0x89EC723: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
    ==111465==    by 0x4A8EAA7: my_raw_malloc(unsigned long, int) (my_malloc.cc:199)
    ==111465==    by 0x4A8E6F0: my_malloc(unsigned int, unsigned long, int) (my_malloc.cc:81)
    ==111465==    by 0x4A8EC74: my_strdup(unsigned int, char const*, int) (my_malloc.cc:296)
    ==111465==    by 0x38A4DE8: wsrep_init_vars() (wsrep_var.cc:51)
    ==111465==    by 0x340D3B1: mysql_init_variables() (mysqld.cc:10643)
    ==111465==    by 0x33FAA5A: init_common_variables() (mysqld.cc:4891)
    ==111465==    by 0x34063E9: mysqld_main(int, char**) (mysqld.cc:7504)
    ==111465==    by 0x31A9919: main (main.cc:25)
    
    Fix: The above memory leaks have been fixed.
-----
2. PXC-3552 - [MTR] Fix main.mysqld--validate-config test (server abort)   
    https://jira.percona.com/browse/PXC-3552
    
    Fixed issues:
    1. When --validate-config option is specified and the datadir is not
       specified, mysqld was trying to change directory and was erroring out
       with 'No such file or directory'. As part of the fix, we skip the
       operation if either --help or --validate-option is specified
    
    2. Server logged PXC warnings even when --help/--validate-config option
       was specified. This is fixed now.
    
    3. Specifying --validate-config created binlog index files. This is
       fixed now.
-----
3. PXC-3553 - [MTR] Fix main.mysqlpump_extended test
    https://jira.percona.com/browse/PXC-3553
    
    Modified the test case to add the PXC users `mysql.pxc.sst.role` and
    `mysql.pxc.internal.session` to the `--exclude-users` mysqlpump option
    to exclude these users while taking the backup.

Testing Done
----------
1. Jenkins: https://pxc.cd.percona.com/view/PXC%208.0/job/pxc-8.0-param/160/testReport/ [In Progress]

Note to reviewers
------
The extra changes in `sql/mysqld.cc` are side-effects of running clang-format-10.

kamil-holubicki commented on sql/mysqld.cc:
```diff
@@ -13027,42 +13027,31 @@ PSI_stage_info *all_server_stages[] = {
     // log_event
-    &stage_wsrep_writing_rows,
-    &stage_wsrep_deleting_rows,
+    &stage_wsrep_writing_rows, &stage_wsrep_deleting_rows,
```
This reformatting will cause the failure of rpl_perfschema_threads_processlist_status.test

venkatesh-prasad-v replied:
Fixed. Disabled `clang-format` check for this block and the test passes now.
```

==============================================================================
                  TEST NAME                       RESULT  TIME (ms) COMMENT
------------------------------------------------------------------------------
[ 20%] rpl.rpl_perfschema_threads_processlist_status 'mix'  [ skipped ]  Doesn't support --binlog-format = 'mixed'
[ 40%] rpl.rpl_perfschema_threads_processlist_status 'row'  [ pass ]    680
[ 60%] rpl.rpl_perfschema_threads_processlist_status 'stmt'  [ skipped ]  Doesn't support --binlog-format = 'statement'
[ 80%] rpl.rpl_perfschema_threads_processlist_status 'ENV'  [ pass ]    707
[100%] shutdown_report                           [ pass ]       
------------------------------------------------------------------------------
The servers were restarted 1 times
The servers were reinitialized 0 times
Spent 1.387 of 30 seconds executing testcases
```

Commit:
PXC-3553 - [MTR] Fix main.mysqlpump_extended test

https://jira.percona.com/browse/PXC-3553

Modified the test case to add the PXC users `mysql.pxc.sst.role` and
`mysql.pxc.internal.session` to the `--exclude-users` mysqlpump option
to exclude these users while taking the backup.

Commit:
PXC-3552 - [MTR] Fix main.mysqld--validate-config test (server abort)

https://jira.percona.com/browse/PXC-3552

Fixed issues:
1. When --validate-config option is specified and the datadir is not
   specified, mysqld was trying to change directory and was erroring out
   with 'No such file or directory'. As part of the fix, we skip the
   operation if either --help or --validate-option is specified

2. Server logged PXC warnings even when --help/--validate-config option
   was specified. This is fixed now.

3. Specifying --validate-config created binlog index files. This is
   fixed now.

Commit:
PXC-3446 - Memory leak during server shutdown

https://jira.percona.com/browse/PXC-3446

Problem: Almost every MTR test exposed the following memory leaks when
run in cluster mode.

1. Memory leak in rollbacker thread arguments.

==111465== 16 bytes in 1 blocks are definitely lost in loss record 2 of 4
==111465==    at 0x89ECC83: operator new(unsigned long)
==111465==    by 0x389DC9A: wsrep_create_rollbacker() (wsrep_thd.cc:309)
==111465==    by 0x3874FBE: wsrep_init_startup(bool) (wsrep_mysqld.cc:1125)
==111465==    by 0x34010C5: init_server_components() (mysqld.cc:6199)
==111465==    by 0x340715B: mysqld_main(int, char**) (mysqld.cc:7689)
==111465==    by 0x31A9919: main (main.cc:25)

2. Memory leak in `wsrep_cluster_name` server variable.

==111465== 49 bytes in 1 blocks are definitely lost in loss record 4 of 4
==111465==    at 0x89EC723: malloc (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==111465==    by 0x4A8EAA7: my_raw_malloc(unsigned long, int) (my_malloc.cc:199)
==111465==    by 0x4A8E6F0: my_malloc(unsigned int, unsigned long, int) (my_malloc.cc:81)
==111465==    by 0x4A8EC74: my_strdup(unsigned int, char const*, int) (my_malloc.cc:296)
==111465==    by 0x38A4DE8: wsrep_init_vars() (wsrep_var.cc:51)
==111465==    by 0x340D3B1: mysql_init_variables() (mysqld.cc:10643)
==111465==    by 0x33FAA5A: init_common_variables() (mysqld.cc:4891)
==111465==    by 0x34063E9: mysqld_main(int, char**) (mysqld.cc:7504)
==111465==    by 0x31A9919: main (main.cc:25)

Fix: The above memory leaks have been fixed.

kamil-holubicki reviewed (approved):
LGTM
